### PR TITLE
core: fix the provisioner graphing

### PR DESCRIPTION
**Pull request by svanharmelen**
Without this change, all provisioners are added to the graph by default and they are never pruned from the graph if their not needed.

I must admit I'm not a 100% sure if this is the correct solution, but there are a few (weird) things that we should discuss and/or have a look at in order to determine the correct solution.

1. I wonder if it makes any sense at all to have a `MissingProvisionerTransformer`? It makes sense to have a `MissingProviderTransformer` because you could have used a resource without specifying the needed `provider` for that resource. But that isn't possible for a `provisioner` as that is a single entity which you either specify in your config, or you don't. So how can you ever have a missing `provisioner` in your config?

2. I think we could ask ourselves the exact same question for the `PruneProvisionerTransformer`. What could this transformer ever prune (if we don't add all available `provisioners` or selves that is, see point 3 :wink:).

3. If we still do want the `MissingProvisionerTransformer`, then I suppose we only want to add the `provisioners` that we actually need. Currently we just add all known `provisioners` and try to prune them in a later stage (which currently doesn't work, see point 4). The code in the PR changes the logic by only adding the `provisioners` that are needed by the supplied config in the same way as this is done for `providers`.

4. We should have a closer look at `PruneProviderTransformer` and `PruneProvisionerTransformer`. They currently both check if there are any `UpEdges`, and only if there none the `provider` or `provisioner` is removed from the graph. But when the `RootTransformer` is executed before one of the prune transformers, they will have at least 1 `UpEdge`, being the `graphNodeRoot` added by the `RootTransformer`. And since the current graph builder [ordering](https://github.com/hashicorp/terraform/blob/master/terraform/graph_builder.go#L103) places the `RootTransformer` before the prune transformers, it means the prune transformers are effectively broken at this moment.

The code in this PR fixes the problem described in point 4 by reordering the transformers, but that doesn't feel like a robust solution. Another (possibly better) fix would be to change [this code](https://github.com/hashicorp/terraform/blob/master/terraform/transform_provisioner.go#L126-L129) to something like:

```go
s := g.UpEdges(v)
if s.Len() == 0 {
  g.Remove(v)
}
if s.Len() == 1 {
  if _, ok := s.List()[0].(graphNodeRoot); ok {
    g.Remove(v)
  }
}
```

This would make the ordering of the graph transformations less critical, but I'm not sure if this is a 100% fix and if this also works when your using modules and/or already have a root, not being a `graphNodeRoot` (not sure that is even possible?). So I need some input to be sure this would indeed be a solid fix for the current prune transformers problem in all use cases...

I labeled this as a bug, because the current code allocates resources (mainly a goroutine with a yamux stream) for each `provisioner` for every action (e.g. `refresh`, `plan`, `apply`) which is never released. If your running Terraform as a long running daemon (API) you can see this will kill your service at a certain point in time :grin:

### CODE REVIEW

**Comment by phinze** [comment]
Hi @svanharmelen, this seems reasonable at first glance!

I'm curious, though, does this have any functional effect?

Also same question for the `RootTransformer` move - is that required to make this change or was that just a cleanup you saw while in there?

**Comment by svanharmelen** [comment]
@phinze I was still typing the complete description for this PR, so please have another look as I suspect it will answer your question.

/CC @mitchellh as this is mainly code he wrote at the time...

**Comment by svanharmelen** [comment]
As for point 2... Maybe the prune provisioner transformer makes sense for when someone specified an empty provisioner block in it's config?

**Comment by svanharmelen** [comment]
@phinze @mitchellh any feedback or thoughts about this one?

**Comment by svanharmelen** [comment]
Any chance we can discuss this one?

If it's up to me I would just remove `MissingProvisionerTransformer` and `PruneProvisionerTransformer` all together and then the only question left is how to properly fix `PruneProviderTransformer`.

The reordering I did in this PR still feel hacky. So would the suggested code change do the trick and be robust enough? Or should we extend that peace of code to check additional use cases/types?

**Comment by mitchellh** [comment]
Looking now, expect a follow up comment shortly.

**Comment by mitchellh** [comment]
My thoughts are: not having the missing and prune transformers make sense. If we can remove those without breaking any tests (other than those that might just test for their existence), then it probably will work fine.

**Comment by svanharmelen** [comment]
Thanks for the feedback! I'll update the PR (somewhere tomorrow) accordingly :+1:

**Comment by svanharmelen** [comment]
@phinze @mitchellh just tried to remove the `MissingProvisionersTransformer`, only to learn that this it is transformer responsible for adding *any* provisioner to the graph. So without this, no provisioners are added to the graph at all :disappointed: 

I first added a few lines of code to the `ProvisionersTransformer` to add any needed provisioners in there. But then I though it would be cleaner to have that logic separated, so I ended up reverting some parts of that :grin: 

To make things a little more descriptive I renamed a few types that had `missing` in their names, as they are not a specific type for missing provisioners, they just represent the (any) provisioners and/or providers in the graph.

So I think the PR should be good to go now, however I do have one feedback question left... I noticed I had to change 1 test in order to make all tests pass: https://github.com/hashicorp/terraform/pull/4877/files#diff-1848efa7a9473a60ae890ae47d35cbf0

I looked at this one, but I fail to fully understand if the previous version is the correct and desired outcome or if this current version is in fact the expected outcome. I was unable to reason about it correctly, so I hope one of you can enlighten me on this one.

Thanks!

File: terraform/transform_provider.go [hunk 1]
```diff
@@ -195,8 +195,8 @@ func (t *MissingProviderTransformer) Transform(g *Graph) error {
 				continue
 			}
 
-			// Add our own missing provider node to the graph
-			m[p] = g.Add(&graphNodeMissingProvider{ProviderNameValue: p})
+			// Add the missing provider node to the graph
+			m[p] = g.Add(&graphNodeProvider{ProviderNameValue: p})
```
Comment (mitchellh) [inline terraform/transform_provider.go]: Why rename this?

**Comment by mitchellh** [comment]
One question about why renaming the missing provider when I believe this PR is about provisioners.

File: terraform/transform_provider.go (hunk 1 above)
Comment (phinze) [inline terraform/transform_provider.go]: I'm actually +1 on this rename or something similar (since this node really does represent the provider itself in the graph) but I'd prefer we do it as a separate PR.

Comment (svanharmelen) [inline terraform/transform_provider.go]: Yeah, that was exactly the argument for renaming this... Since this node represents the provider itself in the graph, this name seemed a little more suitable and more descriptive.

I added the rename in this PR as the provisioner and provider transformers are very much inline with each other and I changed it for the provisioner as well. 

But I'll take this code out if you prefer that and make a new PR for the provider renaming... :+1:

**Comment by svanharmelen** [comment]
Check... Replied inline, but will take out the provider rename and put in it a separate PR.

**Comment by svanharmelen** [comment]
Updated the PR to only contain provisioner related changes...

**Comment by josephholsten** [comment]
@svanharmelen I thought you were pulling out the renames from this branch? do you still need to push those changes?

**Comment by josephholsten** [comment]
ah, all the functional changes are in the first commit. LGTM, though I'm not qualified to vouch for test completeness.

**Comment by phinze** [comment]
Hello again! This is looking good - I believe it solves the `Provisioner already initialized` portion of https://github.com/hashicorp/terraform/issues/3114 - cooking up a test to verify that it does that I'd like to drop in. Will keep you posted.

**Comment by svanharmelen** [comment]
Sounds good... Thx!

**Comment by phinze** [comment]
Okay just confirmed on the linked branch/commit that this does indeed address the provisioner side of that problem. So I'll merge this now and then land that test once the Provider side of the issue is handled (looking into that now).

Thanks for all your work on this, Sander! :+1:

**Comment by svanharmelen** [comment]
Thanks!
